### ISSUE TITLE
Failed to build: attribute rpath is missing

### ISSUE THREAD

**Issue by FrigoEU** [openmoto-org/kontroller#1]
Hey,

I've been trying to get this to build and use in my own rust + esp32 + esp-idf project. The build tooling looks very interesting and much cleaner then mine, but I'm getting an error:

`error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'nix-shell'
         whose name attribute is located at /nix/store/5jgh89kgmrb687c254wxdac4c
j5hqjw8-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'buildInputs' of derivation 'nix-shell'

         at /nix/store/5jgh89kgmrb687c254wxdac4cj5hqjw8-source/pkgs/stdenv/gener
ic/make-derivation.nix:378:7:

          377|       depsHostHost                = elemAt (elemAt dependencies 1
) 0;
          378|       buildInputs                 = elemAt (elemAt dependencies 1
) 1;
             |       ^
          379|       depsTargetTarget            = elemAt (elemAt dependencies 2
) 0;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'rpath' missing

       at /nix/store/kmbd1iacjjbq02d51zjicihy82dh0xq5-source/nix/rust-esp.nix:39
:29:

           38|               --set-interpreter ${stdenv.cc.bintools.dynamicLinke
r} \
           39|               --set-rpath ${lib.rpath} \
             |                             ^
           40|               "$file" || true
       Did you mean path?
`

I'm not sure where this rpath is supposed to come from, but I think it's correct that it's not in nixpkgs.lib. Any ideas?

Simon

**Comment by tshakah** [comment]
Did you get anywhere with this?

**Comment by tshakah** [comment]
I've done some digging and solved it with `rpath = "$out/lib:" + lib.makeLibraryPath [ pkgs.zlib stdenv.cc.cc.lib ];` in a `let` and then `--set-rpath "${rpath}" \` in the two rust*.nix files.

I'm still getting failures when building though:
```
ERROR: This script was called from a virtual environment, can not create a virtual environment again
  Error: Could not install esp-idf

  Caused by:
      command 'env -u IDF_PYTHON_ENV_PATH -u MSYSTEM IDF_TOOLS_PATH="/home/elishahastings/source/seagen/zenoh-pico-rust/.embuild/espressif" "python3" "/home/elishahastings/source/seagen/zenoh-pico-rust/.embuild/espressif/esp-idf/v5.2.2/tools/idf_tools.py" "--idf-path" "/home/elishahastings/source/seagen/zenoh-pico-rust/.embuild/espressif/esp-idf/v5.2.2" "--non-interactive" "install-python-env"' exited with non-zero status code 1
```

That was much further on that this issue

**Comment by ar3s3ru** [comment]
Hi folks, I have completely missed this issue, apologies!

(In my defense, I haven't publicized this repo much, so I was not expecting any kind of traffic or interest in it 👀)

At a high level, I can't tell what the issue could be.
It may depend on what dev environment you're running this in - in my case, it's Macbook M1.

**Comment by ar3s3ru** [comment]
Ah @FrigoEU @tshakah now I understand better where the issue may be coming from.

The contents of the derivation in `nix/rust-esp.nix` (and `rust-src-esp.nix`) are coming from here: https://github.com/nix-community/fenix/blob/111b2aea9d2ac261c6f0a4aac4bde6c0e8c5bb3d/lib/mk-toolchain.nix

`lib` in this scope seems to be coming from `nixpkgs`, delegated to `pkgs.callPackage`: https://github.com/nix-community/fenix/blob/111b2aea9d2ac261c6f0a4aac4bde6c0e8c5bb3d/default.nix#L32

Are you trying to run the devshell in this project, or did some copy-paste of the contents of this project into your own?

(Also, not sure if this is connected, maybe you gotta make sure you are setting up the Rust toolchain like so: https://github.com/openmoto-org/kontroller/blob/7f156946859e7b0502948b6ed30d07f538bdb4c6/flake.nix#L34-L37)

**Comment by tshakah** [comment]
I get the error in the original post if I just clone this repo and `direnv allow`

**Comment by ar3s3ru** [comment]
@tshakah I just cloned the repo on my NixOS machine and I can indeed reproduce the issue.

It should be fixed with https://github.com/openmoto-org/kontroller/commit/b75043361b22dd6ad16a96ea32d3917354a42fb1, please take a look and reopen if that didn't help.

(Note: it might fail again due to missing `sha256` in `rust-esp.nix` -- if you pass me the right value I'll fix it right up 🙏🏻 )

**Comment by FrigoEU** [comment]
Hey @ar3s3ru ,

Nice work! The thing I ran into first was in rust-esp.nix: I added "${stdenv.cc.cc.lib}/lib" to all --set-rpath options, otherwise I ended up with a rustc that could not find "libstdc++.so.6".

After that, I kept running into the same issue: Cargo tries to compile esp-idf-sys, which tries to start a python venv but that clashes with the venv that is exported by mirrexagon/nixpkgs-esp-dev. I've tried a 100 different things but never managed to get past this issue :(.
